modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -416,34 +416,49 @@ private InternalTransaction beginBusy(
 
         if (readOnly) {
             HybridTimestamp beginTimestamp = clockService.now();
-
-            UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp, options.priority());
-
-            tx = beginReadOnlyTransaction(timestampTracker, beginTimestamp, txId, implicit, options);
+            tx = beginReadOnlyTransaction(timestampTracker, beginTimestamp, implicit, options);
         } else {
             HybridTimestamp beginTimestamp = createBeginTimestampWithIncrementRwTxCounter();
-
-            UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp, options.priority());
-
-            // TODO: RW timeouts will be supported in https://issues.apache.org/jira/browse/IGNITE-24244
-            //  long timeout = options.timeoutMillis() == 0 ? txConfig.readWriteTimeout().value() : options.timeoutMillis();
-            long timeout = 3_000;
-
-            tx = new ReadWriteTransactionImpl(this, timestampTracker, txId, localNodeId, implicit, timeout);
+            tx = beginReadWriteTransaction(timestampTracker, beginTimestamp, implicit, options);
         }
 
         txStateVolatileStorage.initialize(tx);
 
         return tx;
     }
 
+    private ReadWriteTransactionImpl beginReadWriteTransaction(
+            HybridTimestampTracker timestampTracker,
+            HybridTimestamp beginTimestamp,
+            boolean implicit,
+            InternalTxOptions options) {
+
+        UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp, options.priority());
```

Comment:
There seems to be quite a lot of duplication with the method that begins an RO transaction

Reply:
Actually, there the duplication is not that huge. It seems to be enough to extract a method to calculate tx timeout (I mean this line `long timeout = options.timeoutMillis() == 0 ? txConfig.readWriteTimeout().value() : options.timeoutMillis();`)

Reply:
Another thing is that 0 is still used to represent `USE_TIMEOUT_DEFAULT_FROM_CONFIG`. When I suggested to add a constant elsewhere, it should have touched `TxManagerImpl` as well, and all other places where we do the same thing; so `ClientTransactions` does not look like the proper place for its introduction.

Reply:
Looks like this constant must be shared between client and server. Do we have such a place to put into? Don't you think it is too much for `0`?

Reply:
This one is a pretty important zero. :) I mean, it is really easy to forget that it means 'take the default from server config' and not actual zero and not 'no timeout at all'. As for where to put it, the `core` module is accessible both by server and client. I don't think it's the *best* place, but it will work for now.

Reply:
I get your point and I believe it is really easy to confuse 'no timeout at all' vs 'use server default' for 0. I've adjusted both places(`TxManagerImpl` and `ClientTransactions`) where we use this 0 and explicitly wrote a java comment. 

But Im not ready to introduce a coupling point here. We do not have shared 'constants' place and I don't think we should.

Reply:
I would not look at it as coupling, it's rather a common vocabulary. But I think this should not hold the PR merge.